Limit expansion in `TemplateConstructableTypes`

MichalStrehovsky opened the pull request:
Fixes #82384

The template type loader is a limited type loader that we use in native AOT to support scenarios such as `MakeGenericType`. We keep around extra MethodTables and metadata to be able to construct new type instantiations (e.g. `List<string>`) at runtime from template instantiations (e.g. `List<__Canon>`) we made at compile time.

The template instantiation is a `MethodTable` like any other that we make a copy of and patch up with the help of the metadata ("native layout metadata"). Patching up involves e.g. building interface list (`List<string>` should have `IList<string>` in the interface list). This patching up may involve loading more new types from templates (e.g. the mentioned `IList<string>` that needs to be loadable from a `IList<__Canon>` template MethodTable).

The job of the compiler is to figure out all the templates we might need (recursively) to build a type. This is done in places using a rather non-exact `TemplateConstructableTypes` call that just decomposes the type and makes templates for _everything_. Some of these might not be actually needed.

This is an attempt to somewhat limit it.

Cc @dotnet/ilc-contrib

MichalStrehovsky committed:
Limit expansion in `TemplateConstructableTypes`

The template type loader is a limited type loader that we use in native AOT to support scenarios such as `MakeGenericType`. We keep around extra MethodTables and metadata to be able to construct new type instantiations (e.g. `List<string>`) at runtime from template instantiations (e.g. `List<__Canon>`) we made at compile time.

The template instantiation is a `MethodTable` like any other that we make a copy of and patch up with the help of the metadata ("native layout metadata"). Patching up involves e.g. building interface list (`List<string>` should have `IList<string>` in the interface list). This patching up may involve loading more new types from templates (e.g. the mentioned `IList<string>` that needs to be loadable from a `IList<__Canon>` template MethodTable).

The job of the compiler is to figure out all the templates we might need (recursively) to build a type. This is done in places using a rather non-exact `TemplateConstructableTypes` call that just decomposes the type and makes templates for _everything_. Some of these might not be actually needed.

This is an attempt to somewhat limit it.

Sergio0694 commented:
Out of curiosity, will all of this be trimmed out if you don't actually use `MakeGenericType`? I'd ask for a feature switch but I imagine you probably wouldn't like that, which is fair. What I'm thinking is: in CsWinRT 3.0 we're going out of our way to avoid doing `MakeGenericType` anywhere. If you have an app that never calls it (but still uses some reflection, like to get attributes), would all of this template type loader infrastructure be trimmed? Does it save a noticeable amount of size?

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
More aggressive

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
Trim interfaces referenced from native layout

MichalStrehovsky committed:
Optimize away GVM resolution metadata

Ever since GVM support was added to native AOT, we were generating the GVM resolution metadata for every type considered allocated. This included GVMs that were never even called (see `TypeGVMEntriesNode` that simply goes over everything on the type). This PR introduces tracking with method level granularity. I ran into this in a different PR where this was dragging `double`/`float` into compilation just because `int` implements generic math.

MichalStrehovsky committed:
Revert "Optimize away GVM resolution metadata"

This reverts commit 66ec27314ee84ff1eab2347ae94744eac14e36b4.

MichalStrehovsky commented:
> The size savings for CsWinRT look not too bad! 👀

This is currently multiple changes in one, I'm peeling some of it off into #118632 for reviewability, but it looks promising!

MichalStrehovsky committed:
Merge branch 'main' into templateconstructable

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky commented:
# Size statistics

Pull request dotnet/runtime#117502

| Project | Size before | Size after | Difference |
|---------|-------------|------------|------------|
| TodosApi-linux | [ 24200488 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764380992) | [ 24155096 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764381038) | -45392 |
| TodosApi-windows | [ 25511936 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764398997) | [ 25396224 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764391551) | -115712 |
| avalonia.app-linux | [ 18865584 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764378856) | [ 18819984 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764385174) | -45600 |
| avalonia.app-windows | [ 19395584 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764386239) | [ 19341824 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764395739) | -53760 |
| hello-linux | [ 1344480 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764380011) | [ 1344480 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764373142) | 0 |
| hello-minimal-linux | [ 1078088 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764373472) | [ 1078088 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764372612) | 0 |
| hello-minimal-windows | [ 824320 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764389822) | [ 823808 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764378102) | -512 |
| hello-windows | [ 1119232 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764385416) | [ 1117696 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764384390) | -1536 |
| kestrel-minimal-linux | [ 5189304 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764374630) | [ 5181032 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764375951) | -8272 |
| kestrel-minimal-windows | [ 4733440 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764391697) | [ 4690432 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764380601) | -43008 |
| reflection-linux | [ 1951824 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764373214) | [ 1935440 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764373117) | -16384 |
| reflection-windows | [ 1811456 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764381031) | [ 1775104 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764380429) | -36352 |
| webapiaot-linux | [ 9670288 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764375425) | [ 9649728 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764380438) | -20560 |
| webapiaot-windows | [ 10228736 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764393901) | [ 10171904 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764395153) | -56832 |
| winrt-component-full-windows | [ 6663680 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764383441) | [ 6606336 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764383860) | -57344 |
| winrt-component-minimal-windows | [ 1690112 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764380306) | [ 1672704 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/16963743070/artifacts/3764389302) | -17408 |

jkotas commented:
The test failure looks weird. Is it related?
```
[FAIL] System.Linq.Expressions.Tests.ExpressionDebuggerTypeProxyTests.ThrowOnNullToCtor(sourceObject: Param_0)
System.AggregateException : One or more errors occurred. (Didn't find DebuggerTypeProxyAttribute on System.Linq.Expressions.PrimitiveParameterExpression`1[System.Int32].) (Object reference not set to an instance of an object.)
---- Microsoft.DotNet.XUnitExtensions.SkipTestException : Didn't find DebuggerTypeProxyAttribute on System.Linq.Expressions.PrimitiveParameterExpression`1[System.Int32].
---- System.NullReferenceException : Object reference not set to an instance of an object.

----- Inner Stack Trace #1 (Microsoft.DotNet.XUnitExtensions.SkipTestException) -----
   at System.Linq.Expressions.Tests.ExpressionDebuggerTypeProxyTests.ThrowOnNullToCtor(Object sourceObject) + 0x205
   at System.Linq.Expressions!<BaseAddress>+0x1df7e62
   at System.Reflection.DynamicInvokeInfo.InvokeWithFewArguments(IntPtr, Byte&, Byte&, Object[], BinderBundle, Boolean) + 0x91
----- Inner Stack Trace #2 (System.NullReferenceException) -----
--- End of stack trace from previous location ---
--- End of stack trace from previous location ---
```

davidwrighton commented:
Looking at this change, it changes the policy of what the expansion analysis is doing.

Before the change, if a type ended up being marked as something that had a template, it would force all the various bits and pieces of the type to be constructed, such that a MakeGenericType on it would succeed reliably.

After the change, the policy is more that template construction will not trigger additional codegen. If it so happens that the needed code for the final application will not work, it is what it is. I'm not currently familiar enough with the code to determine if this will fail in an explainable fashion, but it is plausible that the failures that occur would be explained by the various annotations we have.

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
Throwaway troubleshooting check

MichalStrehovsky committed:
Check we might want to keep

MichalStrehovsky committed:
Merge branch 'main' into templateconstructable

MichalStrehovsky committed:
Handle GC statics

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
One more

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
Merge branch 'main' into templateconstructable
